tikv-client/src/main/java/com/pingcap/tikv/txn/LockResolverClient.java

```diff
@@ -0,0 +1,288 @@
+    }
+
+    List<Lock> expiredLocks = new ArrayList<>();
+    for (Lock lock: locks) {
+      if (TsoUtils.isExpired(lock.getTxnID(), lock.getTtl())) {
+        expiredLocks.add(lock);
+      }
+    }
+
+    if (expiredLocks.size() == 0) {
+      return false;
+    }
+
+    // TxnID -> []Region, record resolved Regions.
+    // TODO: Maybe put it in LockResolver and share by all txns.
+    Map<Long, Set<RegionVerID>> cleanTxns = new HashMap<>();
+    for (Lock l: expiredLocks) {
+      Long status = getTxnStatus(bo, l.getTxnID(), l.getPrimary());
+
+      cleanTxns.putIfAbsent(l.getTxnID(), new HashSet<>());
+      Set<RegionVerID> cleanRegion = cleanTxns.get(l.getTxnID());
+
+      resolveLock(bo, l, status, cleanRegion);
+    }
+
+    return expiredLocks.size() == locks.size();
+  }
+
+  private void resolveLock(BackOffer bo, Lock lock, long txnStatus, Set<RegionVerID> cleanRegion) {
+
+    while (true) {
+      region = session.getRegionManager().getRegionByKey(lock.getKey());
+
+      if (cleanRegion.contains(region.getVerID())) {
+        return ;
+      }
+
+      Supplier<ResolveLockRequest> factory;
+
+      if (txnStatus > 0) {
+        // txn is commited with commitTS txnStatus
+        factory = () ->
+            ResolveLockRequest.newBuilder().setContext(region.getContext()).
+                setStartVersion(lock.getTxnID()).setCommitVersion(txnStatus).build();
+      } else {
+        factory = () ->
+            ResolveLockRequest.newBuilder().setContext(region.getContext()).
+                setStartVersion(lock.getTxnID()).build();
+      }
+
+      KVErrorHandler<ResolveLockResponse> handler =
+          new KVErrorHandler<>(
+              session.getRegionManager(), this, region, resp -> resp.hasRegionError() ? resp.getRegionError() : null);
+
+      ResolveLockResponse resp = callWithRetry(bo, TikvGrpc.METHOD_KV_RESOLVE_LOCK, factory, handler);
+
+      if (resp.hasError()) {
+        logger.error(String.format("unexpected resolveLock err: %s, lock: %s", resp.getError(), lock));
+        throw new KeyException(resp.getError());
+      }
+
+      if (resp.hasRegionError()) {
+        bo.doBackOff(BoRegionMiss, new RegionException(resp.getRegionError()));
+        continue;
+      }
+
+      cleanRegion.add(region.getVerID());
+      return ;
+    }
+  }
+
+  @Override
+  protected TikvBlockingStub getBlockingStub() {
+    return blockingStub.withDeadlineAfter(getConf().getTimeout(), getConf().getTimeoutUnit());
+  }
+
+  @Override
+  protected TikvStub getAsyncStub() {
+    return asyncStub.withDeadlineAfter(getConf().getTimeout(), getConf().getTimeoutUnit());
+  }
+
+  @Override
+  public void close() throws Exception {
+  }
+
+  /**
+   * onNotLeader deals with NotLeaderError and returns whether re-splitting key range is needed
+   *
+   * @param newStore the new store presented by NotLeader Error
+   * @return false when re-split is needed.
+   */
+  @Override
+  public boolean onNotLeader(Store newStore) {
```

Comment:
```
asyncStub
blockingStub
```
and etc can be moved into `AbstractGrpcClient`. In this way, we can also move `onNotLeader` to `AbstractGrpcClient` to avoid duplicated code.

Reply:
I think this can be merged in another PR, because this change will invoke a large change of AbstractGrpcClient